include/forcing/ForcingsEngineDataProvider.hpp

```diff
@@ -0,0 +1,211 @@
+
+    ~ForcingsEngineDataProvider() override = default;
+
+    boost::span<const std::string> get_available_variable_names() override
+    {
+        return var_output_names_;
+    }
+
+    long get_data_start_time() override
+    {
+        return clock_type::to_time_t(time_begin_);
+    }
+
+    long get_data_stop_time() override
+    {
+        return clock_type::to_time_t(time_end_);
+    }
+
+    long record_duration() override
+    {
+        return std::chrono::duration_cast<std::chrono::seconds>(time_step_).count();
+    }
+
+    size_t get_ts_index_for_time(const time_t& epoch_time) override
+    {
+        const auto epoch = clock_type::from_time_t(epoch_time);
+
+        if (epoch < time_begin_ || epoch > time_end_) {
+          throw std::out_of_range{
+            "epoch " + std::to_string(epoch.time_since_epoch().count()) +
+            " out of range of " + std::to_string(time_begin_.time_since_epoch().count()) + ", "
+            + std::to_string(time_end_.time_since_epoch().count())
+          };
+        }
+
+        return (epoch - time_begin_) / time_step_;
+    }
+
+    std::shared_ptr<models::bmi::Bmi_Py_Adapter> model() noexcept
+    {
+        return bmi_;
+    }
+
+    /* Remaining virtual member functions from DataProvider must be implemented
+       by derived classes. */
+
+    data_type get_value(
+      const selection_type& selector,
+      data_access::ReSampleMethod m
+    ) override = 0;
+
+    std::vector<data_type> get_values(
+      const selection_type& selector,
+      data_access::ReSampleMethod m
+    ) override = 0;
+
+  protected:
+    using storage_type = detail::ForcingsEngineStorage;
+
+    ForcingsEngineDataProvider(
+        std::string init,
+        std::size_t time_begin_seconds,
+        std::size_t time_end_seconds
+    )
+      : time_begin_(std::chrono::seconds{time_begin_seconds})
+      , time_end_(std::chrono::seconds{time_end_seconds})
+    {
+        // Get a forcings engine instance if it exists for this initialization file
+        auto bmi_ = storage_type::instances.get(init);
+
+        // If it doesn't exist, create it and assign it to the storage map
+        if (bmi_ == nullptr) {
+            // Outside of this branch, this->bmi_ != nullptr after this
+            bmi_ = std::make_shared<models::bmi::Bmi_Py_Adapter>(
+                "ForcingsEngine",
+                init,
+                forcings_engine_python_classpath,
+                /*allow_exceed_end=*/true,
+                /*has_fixed_time_step=*/true,
+                utils::getStdOut()
+            );
+
+            storage_type::instances.set(init, bmi_);
+        }
+
+        // Now, initialize the BMI dependent instance members
+        // NOTE: using std::lround instead of static_cast will prevent potential UB
+        time_step_ = std::chrono::seconds{std::lround(bmi_->GetTimeStep())};
+        time_current_index_ = (std::chrono::seconds{std::lround(bmi_->GetCurrentTime())} / time_step_); // resolves to long
```

Comment:
`time_current_index_` is initialized and updated, but it's `private` and I never see it being used within this class definition. Is it just a relic of some sort?

Reply:
Yeah I think this was a relic, since I didn't see any reference to it in the gridded or lumped implementations in a previous PR. Removed it, thanks!